[RPC Reliability] Implemented retries for RPCs with exponential backoff

osalpekar opened the pull request:
Summary:
This adds functionality for re-trying RPC's that are sent with the function `sendWithRetries()`. It adds RPC's that will potentially need to be retried to a sorted map that contains the timeout at which to retry the RPC and associated metadata. A separate thread iteratively removes the earliest retry-able RPC from the map, sleeps until the corresponding time point, re-tries the RPC, and adds to the map again with a future timeout.

GitHub Issue: https://github.com/pytorch/pytorch/issues/32124

Per the first milestone, the following will be addressed in future PR's:
* replacing sleeping in RPC Retry Thread with condition variables or timer wheel (current thread sleep will trigger linter)
* specific values for `kRpcRetryDuration`, `maxRetries` default, exact backoff algorithm
* merging `pollTimedOutRpc` thread with RPC Retry Thread
* enabling RPC Retries for RRef internal messages

Differential Revision: D19560159

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit a36f45e0:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=9e0ce72e9e0af2ef21e8c24069c656ab45a18745&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 9e0ce72e since Jan 24
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=a36f45e097fdbe9fdc4331a50c4b0da4d345058f).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=90539420) from Jan 24



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

rohan-varma commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+      profilingEnabled_(false) {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(true);
```
Can we dedupe this with the `std::atomic<bool> rpcRunning_{false};` in processGroupAgent?

osalpekar replied:
Yep, that's the plan. I'll be making `processGroupAgent`-related changes in a future PR per the design doc

rohan-varma replied:
That makes sense. Although, it may be bug prone or at least confusing to see 2 sources of truth for if the RPC agent is running or not, before the PG agent changes get landed, hence why it may be useful to do the deduping here itself. I don't have too much of a preference either way.

osalpekar replied:
Cool, I'll add a comment to the atomic variable in PGA saying it's likely to be replaced with the one in the parent class - hopefully that clears some confusion in the meantime!  I think let's go for doing PGA-stuff in a future PR, since we're also thinking about moving the RPC Timeout Watchdog thread to `RpcAgent` and need to make sure this all plays well with the existing `start()` and `shutdown()` functions. This would also make for more functionally-separated PR's. Open to alternate suggestions though!

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -542,7 +542,19 @@ void ProcessGroupAgent::listenLoop() {
+    try {
+      workCompleted = work->wait();
+    } catch (...) {
+      // gloo timed out waiting for the send to succeed, which throws an
```
nit: won't this be a receive?

osalpekar replied:
The Exception thrown from here is from unbound_buffer.cc in gloo. The error is "Timed out waiting <timeout>ms for send operation to complete"

rohan-varma replied:
That might be the exception on the client end, but this should throw for recv. Above, we call `pg_->recvAnySource()` which returns a `ProcessGroup::RecvWork`. Tracing down the stack we call `UnboundBuffer::waitRecv` in gloo: https://github.com/facebookincubator/gloo/blob/master/gloo/transport/tcp/unbound_buffer.cc#L54, it looks like. Though I don't think it matters too much, we should just catch whatever exception it throws.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -542,7 +542,19 @@ void ProcessGroupAgent::listenLoop() {
+    bool workCompleted;
+    try {
+      workCompleted = work->wait();
+    } catch (...) {
```
What exceptions could we encounter here? it might be better to handle them explicitly and fail when we encounter an unexpected exception (cc @jjlilley who mentioned something similar on a previous diff)

rohan-varma commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+    }
+    steady_clock_time_point earliestTimeout = rpcRetryMap_.begin()->first;
+    if (std::chrono::steady_clock::now() < earliestTimeout) {
+      std::this_thread::sleep_until(earliestTimeout);
```
Could this sleep block the shutdown of the RPC agent?

osalpekar replied:
Yes, the condition variable-based sleeping PR is almost ready (which is Milestone 2). We had decided for the purposes of ensuring the rest of the retry functionality works that we could settle for a simple thread sleep in this first PR.

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+    int maxRetries) {
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+  auto future = std::make_shared<FutureMessage>();
+  auto fm = send(to, std::move(message));
```
you're holding rpcRetryMutex_ too broadly here, e.g. I don't think you should hold across send()

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+      profilingEnabled_(false) {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(true);
```
you should probably just init this true, rather than grabbing the lock here.

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+}
+
+RpcAgent::~RpcAgent() {
+  {
```
This seems like it potentially belongs in the equivalent of shutdown() (I realize this is virtual = 0 now).
It's not awesome practice to block in the destructor like this.

It's probably ok as-is now, but probably want a future change where the subclasses call this super classes's shutdown(), and move to there.

rohan-varma commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+        }
+      });
+
+      // The exponential backoff algorithm being used here is:
```
Could we move this to a helper fucntion?

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -55,6 +55,32 @@ struct TORCH_API WorkerInfo {
+      int maxRetries)
+      : to_(to),
+        message_(message),
+        retryFuture_(retryFuture),
```
std::move() around the shared_ptr<> inits to avoid an extra inc/dec ref

rohan-varma commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+    // successfully completed, or we have re-tried it and scheduled the next
+    // re-try by adding to the rpcRetryMap_. Thus, we can remove the current
+    // entry from the map.
+    rpcRetryMap_.erase(earliestTimeout);
```
If I understood this correctly, it looks like we're calculating a new time, changing the retry future (from `earlistRpc.retryFuture` to `fm` in the case that the retryFuture was not completed, and re-inserting the data back into the map.

Instead of a reinsert + erase, could we just update the entries? We should still have the sorted order we desire because the time has been modified.

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+      LOG(INFO) << "Send try 1 failed\n";
+    } else {
+      // First try succeeded and we can mark the original future as completed.
+      future->markCompleted(message);
```
can we remove early from rpcRetryMap_ in this case?

rohan-varma commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -55,6 +55,32 @@ struct TORCH_API WorkerInfo {
+      const WorkerInfo& to,
+      Message&& message,
+      std::shared_ptr<FutureMessage> retryFuture,
+      std::shared_ptr<FutureMessage> originalFuture,
```
Do we need to take in the shared_ptr by value? This will bump a refcount every time which is ideally overhead that we'd like to avoid.

rohan-varma commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+                std::move(earliestRpc.message_),
+                fm,
+                earliestRpc.originalFuture_,
+                earliestRpc.retryCount_ + 1,
```
Similar to below comment - could we bump retryCount instead of constructing a new entry every time?

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+    }
+    steady_clock_time_point earliestTimeout = rpcRetryMap_.begin()->first;
+    if (std::chrono::steady_clock::now() < earliestTimeout) {
+      std::this_thread::sleep_until(earliestTimeout);
```
pls don't sleep here,  since then the destructor will block for an arbitrary time period.

you should instead use the rpcRetryMapCV_ to do the sleeping above (e.g. calculate a timeout, then cv.wait_for()

rohan-varma commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -93,6 +119,18 @@ class TORCH_API RpcAgent {
       const WorkerInfo& to,
       Message&& message) = 0;
 
+  // At-least once semantics for send. Retries sending the message until an ACK
```
Does ACK in this case mean that the future has been successfully processed and marked as completed? If so we might want to clarify, since ACK can also frequently means that the receval has been acknowledged but the worke has not actually been done

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+          std::chrono::time_point_cast<std::chrono::milliseconds>(
+              std::chrono::steady_clock::now() + timedelta);
+
+      if (earliestRpc.retryCount_ + 1 < earliestRpc.maxRetries_) {
```
can we not update this in place instead?

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
 
+// TODO(osalpekar): Tune RetryDuration to be something in seconds per
+// team recommendations. Kept at 100ms for the sake of Milestone 1.
+const auto kRpcRetryDuration = std::chrono::milliseconds(100);
```
It seems like if we have don't have an end-to-end response in 100ms, we'll immediately retry, right?

If we ever have 100ms of queuing on a server side, or any hiccup, or too much GIL waiting, or  TCP hiccup, we'll start a retry storm, right?

(in addition 1.2 is fairly aggressive as a backoff, maybe 2.0 [or even 1.5] is more appropriate)

I'm wondering if, policy-wise, we might consider:
  - if we get an immediate connection-broken message, we retry right away (which doesn't seem to be the case below)... e.g. if we see TCP FIN on the socket, there's no sense in waiting.
  - having a somewhat longer timeout, if we hear nothing

mrshenli replied:
What if the UDF takes longer to finish or if it is a RRef fetch call?

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+                      const rpc::Message& message,
+                      const c10::optional<utils::FutureError>& futErr) {
+    if (futErr) {
+      LOG(INFO) << "Send try 1 failed\n";
```
We need to differentiate failure conditions which are retry-able, vs. not.

A network connection dropped is very different from an unrecoverable function/level-level error

This might not be completely trivial given how unstructured some of our errors are, but I expect that we want to have a function somewhere to tell us whether a given error is retryable. I suspect we want a whitelist (e.g. connection dropped)

Regardless, there are definitely app-level errors that we want to propagate here, and the retry mechanism needs to interpret ones it doesn't understand as such.

rohan-varma replied:
If we only call `sendWithRetries` on internal messages will this ensure that we're not retrying needlessly? My concern with a whitelist is that we'll have to add all possible network/IO related exceptions and we might miss some and constantly have to update.

jjlilley reviewed (commented):
Thanks for working on this, here are a few initial thoughts:

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+                    << " failed\n";
+        } else {
+          // This try succeeded, so we can mark the original future completed.
+          earliestRpc.originalFuture_->markCompleted(message);
```
Is it possible that this future is marked completed multiple times when the original call or previous retries returns late?

osalpekar replied:
Yes this was possible - I’ll add a check in the else clause so that we only mark the future as Complete if we haven’t marked it complete yet

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+    // exponential backoff) at which this RPC would be re-tried next and add it
+    // to the rpcRetryMap_.
+    if (!earliestRpc.retryFuture_->completed()) {
+      auto fm = send(earliestRpc.to_, std::move(earliestRpc.message_));
```
This `fm` also lives in `futures_` in `ProcessGroupAgent`? Which will never get cleared if the the ack didn't come back?

osalpekar replied:
The un-ACKed RPC’s in the `futures_` map are being cleared if they aren't responded to after `rpcTimeout` by the `futureTimeoutThread_` in ProcessGroupAgent

osalpekar commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
```
Right, I'm moving this down to just before the `rpcRetryMap_.emplace` call below

osalpekar commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+    rpcAgentRunning_.store(false);
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
```
per @jjlilley's comments above, I’ll move the current destructor logic to RpcAgent::shutdown() and call this from the child class shutdown() functions. That way, the retryThread will be shutdown when the PGA/Thrift RPC Agents call shutdown.

osalpekar commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,134 @@ namespace rpc {
+    // exponential backoff) at which this RPC would be re-tried next and add it
+    // to the rpcRetryMap_.
+    if (!earliestRpc.retryFuture_->completed()) {
+      auto fm = send(earliestRpc.to_, std::move(earliestRpc.message_));
```
There’s actually nothing that guarantees that the receiver is still alive - for now if the receiver node is down and doesn’t come back up, we’ll end up sending MAX_RETRIES RPC’s (and then either log it or throw an exception - based on some suggestions around strict/non-strict mode from the design review). Based on @jjlilley's comment above, I can fine tune the retry policy/surface appropriate exceptions based on the underlying errors and , for example, regenerate the gloo pair connection for certain errors, but not sure how I could do this in a backend-agnostic way.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,138 @@ namespace rpc {
+  });
+
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+  rpcRetryMap_[newTime] = std::make_shared<RpcRetryInfo>(
```
nit: you can try using rpcRetryMap_.emplace

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,138 @@ namespace rpc {
+            rpcRetryMap_.erase(newTime);
+          }
+        }
+      });
```
Can move this code to a separate function and dedup this between here and in sendWithRetries.

osalpekar replied:
I was thinking about deduplicating these callbacks as well, but it might end up being a somewhat unpleasant function. Since we’re now adding/moving elements in the map inside the callback, in `sendWithRetries` we must pass the arguments needed to create the struct to the callback (like `to`, `message`, `originalFuture`, `maxRetries`, etc.) since nothing for the current RPC has been added to the map at that point. However, in `retryExpiredRpcs`, we only need to pass in `earliestRpc` and `newTime`. This will result in a callback function with the following signature: `rpcCallbackFunction(newTime, earliestRPC, retryCount, &to, &message, &future, maxRetries)` with everything except newTime being optional and having a default of null. This function would also have 3 layers of nested conditionals to handle whether to add to the map or simply move it around based on the `retryCount`. Can still make this change if that will be cleaner in general!

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -45,6 +167,24 @@ void RpcAgent::enableGILProfiling(bool flag) {
+      std::chrono::steady_clock::now() + timedelta);
+}
+
+bool RpcAgent::shouldRetryRpc(std::shared_ptr<FutureMessage> future) {
```
The RPC should be retried only if the RPC has completed with an error. Note that every future will eventually be marked as completed via one of these mechanisms:

1. The work completed successfully.
2. Something failed and the future was marked with an error.
3. The RPC timed out and the RpcAgent (ProcessGroup/Thrift) marks the Future with a timed out exception.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,138 @@ namespace rpc {
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+    auto& earliestRpc = rpcRetryMap_[earliestTimeout];
```
We should probably use `rpcRetryMap_.begin()` here since an element with an earlier timeout might have been added to the map while we were sleeping.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,138 @@ namespace rpc {
+        // retry time.
+        earliestRpc->retryFuture_ = fm;
+        rpcRetryMap_[newTime] = std::move(earliestRpc);
+        rpcRetryMap_.erase(earliestTimeout);
```
Unless I'm missing something it seems like if we exceed maxRetries we don't mark the original future as failed and we don't remove the future from rpcRetryMap_ either.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -165,6 +206,35 @@ class TORCH_API RpcAgent {
+
+  // Map that stores metadata for RPC's that may need to be re-tried as well as
+  // the timepoint at which we should re-try them.
+  std::map<steady_clock_time_point, std::shared_ptr<RpcRetryInfo>> rpcRetryMap_;
```
This should be a map from time_point to a vector of RpcRetryInfo since we could have RpcRetryInfo for two RPCs with the same timepoint.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,138 @@ namespace rpc {
+    }
+  });
+
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
```
Instead of adding the RpcRetryInfo to the map, we can move this logic to the callback. Basically, if in the future callback we receive an error we add the entry to the map.

The same thing can be done in retryExpiredRpcs where we pick the first RPC to retry, remove it from the map, retry it and in the callback decide to add it back to the map if we haven't exceeded maxRetries. This way we automatically clean up the map and also no longer need `shouldRetryRpc` and `retryFuture_`

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -130,7 +171,7 @@ class TORCH_API RpcAgent {
   // Stop accepting requests and shutdown the RPC framework as soon as possible
   // by terminating all RPC threads.
-  virtual void shutdown() = 0;
+  virtual void shutdown();
```
From an API perspective for RpcAgent, it would be cleaner to introduce a protected pure virtual function called `shutdownImpl` that ProcessGroupAgent and ThriftAgent use to implement their shutdown mechanics. RpcAgent.shutdown() cleans up its state and then just calls `shutdownImpl`.

mrshenli replied:
This no longer needs to be virtual, as implementations would override `shutdownImpl`?

pritamdamania87 reviewed (changes requested):
Requesting changes since I think we need to modify `rpcRetryMap_`. Also a unit test as follows would be pretty useful:

1. Inject error simulation in ProcessGroupAgent.
2. Kick off 1000 sendWithRetries with an infinite number of max retries.
3. Disable error simulation in ProcessGroupAgent.
4. Verify all of the 1000 futures complete successfully.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -542,7 +543,19 @@ void ProcessGroupAgent::listenLoop() {
+    }
+    bool workCompleted;
+    try {
+      workCompleted = work->wait();
```
Doing `rpcRunning_.store(false);` may not be the best approach here because this will mean that `::shutdown()` code will not run which joins threads etc. Is this addressing https://github.com/pytorch/pytorch/issues/29703? If so, it should probably be done in a separate PR. I was working on this issue as well so we should sync on how it can be best addressed.

rohan-varma replied:
(See https://github.com/pytorch/pytorch/pull/32957)

pritamdamania87 replied:
Can we back out the listenLoop changes from this PR and patch this PR on top of https://github.com/pytorch/pytorch/pull/32957? Its easier to review all the listenLoop changes in one place.

osalpekar commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -199,6 +199,8 @@ class ProcessGroupAgent : public RpcAgent {
   // should be running.
   // We lock access to this in shutdown() and pollTimedOutRPCs() to prevent race
   // conditions when notifying condition variables.
+  // TODO: This will likely be merged with a similar atomic variable in the
```
Here is the issue: https://github.com/pytorch/pytorch/issues/32963

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,171 @@ namespace rpc {
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
```
nit: validate that this is non-negative.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,171 @@ namespace rpc {
+            .append(").\n");
+        future->setError(errorMessage);
+      }
+    } else if (!future->completed()) {
```
Why do we need this check? Is it possible the future has already been marked as completed by some other mechanism?

osalpekar replied:
If another send attempt was completed successfully and already marked the future as completed, then we shouldn't do anything. The `future::markCompleted()` has a TORCH_CHECK that throws if the future has already been completed, so this will guard against that. I believe @mrshenli suggested this might be a good check in earlier review.

pritamdamania87 replied:
If another send attempt marked the future as completed successfully, then we wouldn't continue retries and wouldn't reach here. IIUC, it looks like the future can be marked completed only once since we only have one outstanding send at a time. Once a send() function completes (successfully/unsuccessfully), only then we kick off a retry.

osalpekar replied:
Ah yes you're right - the previous case was only possible back when the map wasn't manipulated through the callback. I'll change this per your suggestion.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,171 @@ namespace rpc {
+        if (futErr) {
+          LOG(INFO) << "Send try "
+                    << std::to_string(earliestRpc->retryCount_ + 1)
+                    << " failed\n";
```
nit: you don't need the trailing "\n" in LOG(..) messages.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,171 @@ namespace rpc {
+      auto& earliestRpc = *it;
+      steady_clock_time_point newTime =
+          computeNewRpcRetryTime(earliestRpc->retryCount_);
+      auto fm = send(earliestRpc->to_, std::move(earliestRpc->message_));
```
Shouldn't we do `earliestRpc_->retryCount_++` here instead of on line 132?

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,171 @@ namespace rpc {
+                          const c10::optional<utils::FutureError>& futErr) {
+        if (futErr) {
+          LOG(INFO) << "Send try "
+                    << std::to_string(earliestRpc->retryCount_ + 1)
```
This should should be retryCount_ if we update the retryCount_ above.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,171 @@ namespace rpc {
+                .append(
+                    "The RPC has not succeeded after the specified number of max retries (")
+                .append(std::to_string(earliestRpc->maxRetries_))
+                .append(").\n");
```
Use `c10::str(` instead to build the string here. Also, note that you don't need the trailing \n

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,171 @@ namespace rpc {
+                .append(").\n");
+            earliestRpc->originalFuture_->setError(errorMessage);
+          }
+        } else if (!earliestRpc->originalFuture_->completed()) {
```
Same as above, is there any reason the future might have been marked completed by any other mechanism? If not, we should make this just an else statement.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,171 @@ namespace rpc {
+  Message& msg = message;
+  auto fm = send(to, std::move(message));
+
+  fm->addCallback([this, future, newTime, &to, &msg, maxRetries](
```
It seems like we can combine this callback with the one below by capturing RpcRetryInfo and newTime. For the case here, we already have to, message and originalFuture. For retryCount_ we can set it to 1 and maxRetries remains the same.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -93,6 +119,21 @@ class TORCH_API RpcAgent {
+  //
+  // Sends ``message`` to the ``RpcAgent`` of id ``to`` and returns a
+  // ``FutureMessage`` ptr, just like send(). Caller can specify the maximum
+  // number of retries for this RPC (default is 10). This API might end up
```
default is actually 3 in code?

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,160 @@ namespace rpc {
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  TORCH_CHECK(maxRetries >= 1, "maxRetries cannot be negative.");
```
should be >= 0?

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,160 @@ namespace rpc {
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
```
Just realized that in the code this is used as maxTries and not maxRetries. For example if maxRetries is 3, we should try a total of 4 RPCs (the original RPC and then 3 retries). Can we update the code to follow maxRetries appropriately?

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,160 @@ namespace rpc {
+      if (retryCount == 1) {
+        // The retry thread waits for the map to be populated. Thus we notify
+        // once an item has been added the first time.
+        rpcRetryMapCV_.notify_one();
```
Why do we have this specific check? Why can't we just notify always after we emplace in rpcRetryMap_ above?

pritamdamania87 commented:
Changes look good, although there are a bunch of CI failures internally and externally, can you fix those?

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,159 @@ namespace rpc {
-RpcAgent::~RpcAgent() = default;
+    if (!rpcRetryMap_.empty()) {
+      earliestTimeout = rpcRetryMap_.begin()->first;
+      rpcRetryMapCV_.wait_until(lock, earliestTimeout, [&, this] {
```
It seems like we should be able to combine this wait_until with the previous wait, with sufficient lambda condition.

mrshenli replied:
@jjlilley  could you please elaborate on how these two can be combined? Looks like they can be put into a if-else clause, so that we only need to wait once, but not sure if we can use a single wait/wait_until to achieve that.

jjlilley replied:
@mrshenli Sure- it's probably easier to express with the traditional, non-lambda calling convention-

I think something like the following would unify the condvar waiting (maybe I'm missing something though?)

   for (;;) {
      if (!rpcAgentRunning_.load()) break;
       auto now = std::chrono::steady_clock::now();
       auto earliestTimeout = rpcRetryMap_.empty() ? now + kManyYears : rpcRetryMap_.begin()->first;
       if (now >= earliestTimeout) break;
       rpcRetryMapCV_.wait_until(lock, earliestTimeout);
    }

IMO, unifying the waits tends to make the code easier to maintain over time, though there could be other considerations here.

osalpekar replied:
@jjlilley I tried that snippet, and played around with a few more potential solutions for unifying the condvar waiting as much as possible. The above snippet works if we set an arbitrarily large duration for `kManyYears`, for example `std::chrono::hours(10000)`, but the tests time out when using `std::chrono::time_point::max()`. I can drill down on this further and create an issue to track, if using a large constant for the wait duration for the empty map is not preferred. An alternative is as follows:

```
auto earliestTimeout = std::chrono::time_point<std::chrono::steady_clock>::max();
for (;;) {
  if (!rpcAgentRunning_.load()) break;
  if (std::chrono::steady_clock::now() >= earliestTimeout) break;
  if (rpcRetryMap_.empty()) {
    rpcRetryMapCV_.wait(lock);
  } else {
    earliestTimeout = rpcRetryMap_.begin()->first;
    rpcRetryMapCV_.wait_until(lock, earliestTimeout);
  }
}
```
This doesn't consolidate into a single condvar wait, however. I'm happy to go with whichever one you feel is better coding standards-wise!

osalpekar replied:
According to (https://stackoverflow.com/questions/27726818/stdcondition-variablewait-for-exits-immediately-when-given-stdchronodura), using the max time_point with condvar wait_until is a known issue. Seems like the implementation in `ProcessGroupAgent::pollTimedOutRPCs()` on line 528 does `wait_until` if the expiration time is not max, otherwise it uses `wait`.

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,159 @@ namespace rpc {
+      fm->addCallback([this, newTime, earliestRpc](
+                          const rpc::Message& message,
+                          const c10::optional<utils::FutureError>& futErr) {
+        rpcRetryCallback(message, futErr, newTime, earliestRpc);
```
Don't hold the lock across this addCallback() (I'd also say that about send(), but slightly less critical there).
This is inviting a deadlock on an insta-fail.
Typically, you make a list, then do after releasing the lock.

mrshenli replied:
good catch! The callback here might launch right away if the fm has been marked complete, which would lead to deadlock as it will try to acquire `rpcRetryMutex_` again.

osalpekar commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,159 @@ namespace rpc {
+    // just the retry count.
+    LOG(INFO) << "Send try " << std::to_string(earliestRpc->retryCount_ + 1)
+              << " failed";
+    if (earliestRpc->retryCount_ < earliestRpc->maxRetries_) {
```
We essentially decided all non-application-layer errors are retryable, and since `sendWithRetries` will only be used for RRef internal messages, application-level errors are bugs we should fix. We're taking a closer look at this in a separate PR (#32957), and so I've removed the try-catch from process_group_agent.cpp entirely in order to keep all error-catching related changes in the other PR.

mrshenli reviewed (commented):
Do you know why `test_duplicate_name` is failing?

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,159 @@ namespace rpc {
+  Message& msg = message;
+  auto fm = send(to, std::move(message));
+  auto firstRetryRpc = std::make_shared<RpcRetryInfo>(
+      to, std::move(msg), originalFuture, /* retryCount */ 0, maxRetries);
```
`msg` here is a reference to a **moved** `Message`? The move in `send` should have already cleared the payload in `message`. Is this intentional?

jjlilley replied:
btw, I think this line ends up being correct if line 49 is "Message msg = message;" rather than "Message& msg = message;"

ideally add a short comment, and use variable names, to make the copy explicit, to avoid future regressions, e.g.

    Message msgCopy = message; // make a copy of the original for retries
    auto fm = send(to, std::move(message));
   auto firstRetryRpc = std::make_shared<RpcRetryInfo>(
      to, std::move(msgCopy), originalFuture, /* retryCount */ 0, maxRetries);

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -55,6 +58,29 @@ struct TORCH_API WorkerInfo {
+struct TORCH_API RpcRetryInfo {
+  RpcRetryInfo(
+      const WorkerInfo& to,
+      Message&& message,
```
Does this mean we actually need multiple copies of the message? Every time we retry send, we need to pass a rvalue message to the `send` API. Is there a way to just use one copy of the message?

jjlilley replied:
btw, I think any send-with-retry api in general is going to require at least 1 copy per try (including the original try).

Tensor copies should generally be cheap due to refcounting (maybe I'm missing a case?).
Payload will have to be copied, though typically our Message sizes are dominated by tensors, so might not be as bad as it sounds.

It's good that you called out the double std::move() calls here, though - those would have been unfortunate (and probably mean that we probably could have better test coverage of this new api).

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,159 @@ namespace rpc {
+      auto& earliestRpc = *it;
+      steady_clock_time_point newTime =
+          computeNewRpcRetryTime(earliestRpc->retryCount_);
+      auto fm = send(earliestRpc->to_, std::move(earliestRpc->message_));
```
This will move the message multiple times?

osalpekar replied:
I'll make the change that @jjlilley suggested above. Essentially I'll make a copy of the message and move that copy in the `send()` call.

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,159 @@ namespace rpc {
+      // scheduling it for a future retry.)
+      {
+        std::lock_guard<std::mutex> retryMapLock(rpcRetryMutex_);
+        rpcRetryMap_[newTime].emplace(std::move(earliestRpc));
```
I am trying to understand the logic here. The retry is only inserted when the previous future is either timeout or failed, so the retry delay is actually `rpc timeout + backoff`, right? And this is also why we don't need to worry about marking a future as complete multiple times?

osalpekar replied:
Right, since we moved all scheduling of RPC's to the callbacks, there is only one outstanding send for a given RPC at any given time, so there's no need to check whether the original future has already been marked complete. This change also causes the effective retry duration to be `rpc timeout + backoff` as you mentioned, since we wait for the previous send to fail before the callback schedules the next try.

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -160,11 +201,55 @@ class TORCH_API RpcAgent {
+
+  // Function that uses the exponential backoff algorithm to compute the next
+  // time point to retry a given RPC.
+  steady_clock_time_point computeNewRpcRetryTime(int retryCount);
```
should this be inline method?

mrshenli commented:
lint failure looks real:

```
  {
    path: 'torch/csrc/distributed/rpc/rpc_agent.cpp',
    start_line: 14,
    end_line: 14,
    start_column: 1,
    end_column: 1,
    annotation_level: 'failure',
    message: '[cppcoreguidelines-pro-type-member-init] warning: constructor does not initialize these fields: rpcAgentRunning_'
  }
```

osalpekar commented:
All of @mrshenli , @pritamdamania87 , and @jjlilley 's should be incorporated now - thanks for the reviews!

mrshenli reviewed (approved):
Thanks a lot for working on this and being patient on the long review discussion. :)

LGTM! Stamp to unblock, but let's address lint failure and wait for tests before landing. As `sendWithRetry` is not enabled in the code path yet, I think it's OK to let this in, and then we can test and benchmark it in followup PRs. What's the plan for testing? Do we need to implement a faulty rpc agent to test retry?


```
  {
    path: 'torch/csrc/distributed/rpc/rpc_agent.cpp',
    start_line: 123,
    end_line: 123,
    start_column: 15,
    end_column: 15,
    annotation_level: 'failure',
    message: '[performance-for-range-copy] warning: loop variable is copied but only used as const reference; consider making it a const reference'
  }
```

osalpekar committed:
[RPC Reliability] [WIP] Implemented retries for RPCs with exponential backoff (#32602)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32602

This adds functionality for re-trying RPC's that are sent with the function `sendWithRetries()`. It adds RPC's that will potentially need to be retried to a sorted map that contains the timeout at which to retry the RPC and associated metadata. A separate thread iteratively removes the earliest retry-able RPC from the map, sleeps until the corresponding time point, re-tries the RPC, and adds to the map again with a future timeout.

GitHub Issue: https://github.com/pytorch/pytorch/issues/32124

Per the first 3 milestones, the following will be addressed in future PR's:
* enabling RPC Retries for RRef internal messages

Differential Revision: D19560159

fbshipit-source-id: 601ef11d6ad089708e065fb1a0450d3d15a3950a

osalpekar commented:
> Thanks a lot for working on this and being patient on the long review discussion. :)
> 
> LGTM! Stamp to unblock, but let's address lint failure and wait for tests before landing. As `sendWithRetry` is not enabled in the code path yet, I think it's OK to let this in, and then we can test and benchmark it in followup PRs. What's the plan for testing? Do we need to implement a faulty rpc agent to test retry?
> 
> ```
>   {
>     path: 'torch/csrc/distributed/rpc/rpc_agent.cpp',
>     start_line: 123,
>     end_line: 123,
>     start_column: 15,
>     end_column: 15,
>     annotation_level: 'failure',
>     message: '[performance-for-range-copy] warning: loop variable is copied but only used as const reference; consider making it a const reference'
>   }
> ```

Thanks @mrshenli , just about to push the lint fix, and there's some testing code on Phabricator!

rohan-varma commented:
Do we know why `test_duplicate_names` is failing on this PR?

jjlilley commented:
fwiw, the change LGTM as well (thanks for working on this!!)

Per Rohan's comment, can we not land without looking into test_duplicate_names, though? Sometimes OSS tests are flaky, but the 7 broken targets seem to imply something more

zhangguanheng66 commented:
The master branch is broken and I have to revert this PR.
https://circleci.com/gh/pytorch/pytorch/4501086?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

mrshenli commented:
Sorry, my bad, I should have waited for the test before stamping it. Thanks @zhangguanheng66 for reverting.
